rollup/core/classes/RollupCalculator.cls

```diff
@@ -1258,6 +1271,177 @@ public without sharing abstract class RollupCalculator {
+      this.setupDelimiters(opFieldOnLookupObject);
+    }
+
+    public String getGroupingColumn(String fieldName, Object groupingValue, Boolean isFirst, Boolean isLast) {
+      return String.valueOf(groupingValue) + this.delimiter;
+    }
+
+    public String getGroupingStart() {
+      return '';
+    }
+
+    public String getGroupingRowStart(String groupingKey) {
+      return this.metadata.GroupByRowStartDelimiter__c + groupingKey;
+    }
+
+    public String getGroupingRowEnd() {
+      return this.metadata.GroupByRowEndDelimiter__c;
+    }
+
+    public String getEnd(String groupByString) {
+      return groupByString.removeEnd(this.getGroupingRowEnd());
+    }
+
+    private void setupDelimiters(Schema.SObjectField opFieldOnLookupObject) {
+      if (this.metadata.GroupByRowEndDelimiter__c == null) {
+        // insane hack to identify rich text fields
+        Boolean isRichText = JSON.serialize(opFieldOnLookupObject.getDescribe()).contains('"extraTypeInfo":"richtextarea"');
+        this.metadata.GroupByRowEndDelimiter__c = isRichText ? '<br>' : '\n';
+      }
+
+      if (this.metadata.GroupByRowStartDelimiter__c == null) {
+        this.metadata.GroupByRowStartDelimiter__c = '• ';
+      }
+    }
+  }
+
+  private class TableGroupingFormatter implements GroupingFormatter {
+    private final Map<String, Schema.SObjectField> fieldNameToToken;
+    private final Schema.SObjectField targetField;
+    private final Set<String> columnHeaders = new Set<String>();
+
+    private final String tdOpen = '<td>';
+    private final String tdClose = '</td>';
+    private final String thOpen = '<th>';
+    private final String thClose = '</th>';
+    private final String trOpen = '<tr>';
+    private final String trClose = '</tr>';
+
+    public TableGroupingFormatter(Schema.SObjectField targetField) {
+      this.fieldNameToToken = targetField.getDescribe().getSObjectType().getDescribe().fields.getMap();
+      this.targetField = targetField;
+    }
+
+    public String getGroupingColumn(String fieldName, Object groupingValue, Boolean isFirst, Boolean isLast) {
+      String prefix = isFirst ? this.trOpen : '';
+      String suffix = isLast ? this.trClose : '';
+      this.columnHeaders.add(prefix + this.thOpen + this.fieldNameToToken.get(fieldName).getDescribe().getLabel() + this.thClose);
+      if (isLast) {
+        this.columnHeaders.add(this.thOpen + this.targetField.getDescribe().getName() + this.thClose + suffix);
+      }
+      return prefix + this.tdOpen + String.valueOf(groupingValue) + this.tdClose + suffix;
+    }
+
+    public String getGroupingStart() {
+      return '<table>' + String.join(new List<String>(this.columnHeaders), '');
+    }
+
+    public String getGroupingRowStart(String groupingKey) {
+      return groupingKey.replace(this.trClose, '') + this.tdOpen;
+    }
+
+    public String getGroupingRowEnd() {
+      return this.tdClose + this.trClose;
+    }
+
+    public String getEnd(String groupByString) {
```

Comment:
Not a big deal either way, but based on the other method names, it feels like the method name `getEnd()` doesn't quite follow the convention use for the other method names. Something like `getGroupingEnd()` would be a more consistent name, which would give you these 4 methods:
1. `getGroupingStart()`
2. `getGroupingRowStart(String groupingKey)`
3. `getGroupingRowEnd()`
4. `getGroupingEnd(String groupByString)` (formerly `getEnd()`)

Reply:
That's a great callout, thank you!